### ISSUE TITLE
Experiments on KILT-ELI5 dataset

### ISSUE THREAD

**Issue by jdpsen** [primeqa/primeqa#135]
_jaydesen created the following on Feb 01:_  

 Plan is to use AS2 for selecting a diverse set of top ranked sentences and then train a generative pipeline using ELI5 data.

An example from ELI5 dataset:
<img width="553" alt="Screenshot 2022-02-01 at 6 50 01 PM" src="https://media.github.ibm.com/user/7124/files/d2181b80-838f-11ec-96da-9060b7d5d1c9">
 

_Original issue: https://github.ibm.com/ai-foundation/PrimeQA/issues/20_

**Comment by jdpsen** [comment]
_On Feb 07, mabornea commented:_
Feb 8th, 2022

1. Information retrieval 
- Generate training data to train a retriever and/or reranker
- Start with a BM25 index of Wikipedia passages
- Queries are a concatenation of question and answer. Only include answers with at least rating of at least 3. Discard questions with no answers with at least 3 rating
- Rerank the retrieved passages using Colbert trained on NQ or a BERT classifier based reranker 
- Select top N passages as positives and select negative using some heuristics or use in-batch negatives to fine-tune a ColBERT retriever
- the retriever is used to obtain context passages for the answer generator
- an assumption is that there are useful passages in Wikipedia

2. Analysis of the retrieved passages (BM25): 
 - We saw that using the Q + A  to retrieve passages is beneficial as opposed to using the Q only 
 - Returned passages are diverse already and it does not seem that we should worry about this metric yet
 - Not all passages are relevant. Good passages do not have a high lexical overlap with the question itself. The relevant information is in the neighborhood of the question terms in the retrieval corpus.

 - Basic idea: A re-ranking system, Passage Answer Selection (PAS) inspired by AS2:
   1. Use NQ and other open QA dataset to pre-train PAS
   2. Fine-tune with question answer pairs from ELI 5
   3. Use PAS to filter the BM25 passages for training neural IR and rank the passage seen by the generator





3. Closed-Book Reader
 - We trained a baseline reader that generates an answer given only the question, w/o looking at any passages. We used greedy decoding for generation. In the table below, we compare its performance on the ELI5 dev set with results reported in the "Hurdles to..." paper (the system of row 1 is currently the 2nd best on the KILT ELI5 leaderboard):

| Reader  | ROUGE-L |
| ------------- | ------------- |
| Retrieved Passages (paper)  | 24.42  |
| Random Passages (paper)  | 24.20 |
| No Passage (ours)  | 24.44 |

**Comment by jdpsen** [comment]
_On Feb 15, mrglass commented:_
Under /dccstor/few-shot-rel/eli5/predictions I added prediction files for bm25, dpr and reranker models. The best initial retrieval results are under dpr as nq_re2g_dev.jsonl. While the best reranker result is qabase_reranker_re2g_nq_re2g_dev.jsonl

**Comment by jdpsen** [comment]
_On Feb 15, mabornea commented:_
Using the provenance spans as input to the generator does not improve performance.
The dev set for KILT ELI-5 was split into 2 parts: 75% train, 25% validation. The validation set is 377 questions.  
Two reader models: Using the question only as input and using question + provenance spans

```
model1: question only
'f1': 0.24570993939127034,
'rougel': 0.23040764066487857},
```

```
model2: question + concatenated provenance spans
'f1': 0.24842884339235127,
'rougel': 0.2300689244009291},
```

**Comment by jdpsen** [comment]
_On Feb 22, jaydesen commented:_
[ new issue for better  metric  - based on claim] 
Subset the dataset for "meaningful" evaluation.

**Comment by jdpsen** [comment]
_On Feb 22, jaydesen commented:_
Retrieve+ rerank+ generator - needs evaluation. only retriever+generator doesn't give good results.
Provenance did not help - not really exhaustive , only q was considered.

**Comment by jdpsen** [comment]
_On Mar 01, jaydesen commented:_
Some questions are very hard to answer- need a proper answerable subset
Retrieval- Bhavani 
Reranker- Arafat : trained with (1) NQ sentence level - has answer yes/no (2) CNN summary dataset  (given) Article->summary, use summary sentences as query and mark the text para from the article as passage  (some noise) , extension QG ON summary sentence

**Comment by jdpsen** [comment]
_On Mar 14, jaydesen commented:_
@mabornea @Arafat-Sultan : Can you put the numbers we have been discussing for reranker training results here.

**Comment by jdpsen** [comment]
_On Mar 14, mabornea commented:_
Performance of the reranker.  Models trained on just ASNQ or CNN-DM do not have good transfer performance. The best model is the one trained on a combination of ASNQ and CNN-DM. 

| Re-Ranker Training  | Test           | MRR   | R@3 | R@5|
| -------------------- |-----------|-------|------|------|
| ASNQ                         | CNN-DM |  39.8 | 44.0  | 60.8 | 
| CNN-DM                    | ASNQ      | 42.6  | 48.8 | 61.2  |
| ASNQ + CNN-DM     |  ASNQ     |  84.2 | 92.6. | 96.8  |
| ASNQ + CNN-DM     | CNN-DM | 95.2. | 98.7  | 99.8  |
| ASNQ + CNN-DM-Q | ASNQ      | 88.2  | 94.4  | 96.9  |
| ASNQ + CNN-DM-Q | CNN-DM | 91.6  | 96.7  | 99.0 |

Initial inspection of ELI5 reranked passages shows some improvement.

**Comment by jdpsen** [comment]
_On Mar 15, jaydesen commented:_
ASNQ + CNN-DM-Q models applied on eli5 subset (sample subset) , compared to bm25 it shows potential improvements after reranking.
Use this to train a neural IR.

**Comment by jdpsen** [comment]
_On Mar 24, Arafat-Sultan commented:_
Closed-book QA results on the ELI5 official dev set:

| Training Examples  | # of Examples            | F1   | ROUGE-L |
| -------------------- |-----------|-------|------|
| Annotation Score >= Median (3) | 526k | 25.7 | 24.4 |
| Most Similar Answer Pairs (w/ some filtering) | 165k | 27.2 | 25.9 |

These results indicate that the similar answer pairs heuristic is useful.

**Comment by jdpsen** [comment]
_On Mar 24, jaydesen commented:_
KILT ELI 5 train set is not processed. - not getting god retrieval docs from wiki. Question subset to have a better train set that can answer from Wikipedia. 
CC -C4 version might give better retriever provenance for ELI5.

**Comment by jdpsen** [comment]
_On Mar 31, jaydesen commented:_
Re-ranker trained on CNN-Q,ASNQ, ELI5 (subset selected for training) works well as a standalone model. Plan is to complete the Wikipedia exp.  before moving to CC. Bhavani to produce retrieval results with a revised heuristic-based expansion of answer words: include more from answer words if they appear in any of the other answers 
@Arafat-Sultan : please update the numbers you were showing last day.

**Comment by jdpsen** [comment]
_On Mar 31, Arafat-Sultan commented:_
Re-ranking results of the model jointly trained on CNN-DM, ASNQ and ELI5:

CNN-DM:
| MRR | MR@1 | MR@2 | MR@3 | MR@5 |
| ----------- |-----------|-----------|-----------|-----------|
| 92.8 | 88.2 | 94.4 | 97.4 | 99.2 |

ASNQ:
| MRR | MR@1 | MR@2 | MR@3 | MR@5 |
| ----------- |-----------|-----------|-----------|-----------|
| 89.5 | 83.1 | 92.1 | 95.4 | 97.4 |

ELI5:
| MRR | MR@1 | MR@2 | MR@3 | MR@5 |
| ----------- |-----------|-----------|-----------|-----------|
| 94.9 | 90.7 | 97.7 | 99.6 | 99.9 |

**Comment by jdpsen** [comment]
_On Apr 05, jaydesen commented:_
Bhavani: additonal heuristics on BM25 to utilize the answer. 
To-retry: the retrieval on Wikipedia. 
Sara reranked Wikipedia passages based on answer similarity. Restrict the training set to questions which has Wikipedia passages similar to derived similar answers (from Arafat).

**Comment by jdpsen** [comment]
_On Apr 05, sjrosenthal commented:_
Similarity of Wikipedia passages and ELI5 answers

Total 5000 train-dev
Pass threshold > .1
similarity from ELI5 only: 1543
similarity from ELI5 + Wiki: 1948 

 answer pair score |  >.1 and   <=.3 |  > .3 | 3rd answer >.3 (subset of   prior column)
-- | -- | -- | --
eli5 - eli5 | 183 | 26 |  NA
eli5 - wiki | 49 | 1 |  NA
eli5 - wiki - eli5 | 117 | 64 | 35
eli5 - wiki - wiki | 434 | 234 | 161
eli5 - eli5 - eli5 | 199 | 210 | 108
eli5 - eli5 - wiki | 224 | 207 | 78


.25 Appears to be an appropriate threshold

**Comment by jdpsen** [comment]
@mihaela-bornea @arafat: is this being worked on actively ? Please put your updates and pipeline label accordingly.

**Comment by mihaela-bornea** [comment]
We moved from transformers 4.3.3  to transformers 4.17.0.
With the new transformers version, previous training selection strategy did not show improvement over the baseline. 
Not active at the moment.
